### PR TITLE
[CALCITE-3195] Handle UDF that throws checked exceptions in enumerable code generator

### PR DESCRIPTION
**Pull request by DonnyZone**
Details are described in [JIRA](https://issues.apache.org/jira/browse/CALCITE-3195).
This PR handles checked exceptions declared in method.
(1) Wrap a method call in a `try...catch...` block;
(2) Enable BlockBuilder to inline variables in `TryStatement`.

### CODE REVIEW

File: linq4j/src/main/java/org/apache/calcite/linq4j/tree/TryStatement.java [hunk 1]
```diff
@@ -36,7 +37,17 @@ public TryStatement(Statement body, List<CatchBlock> catchBlocks,
...
+    shuttle = shuttle.preVisit(this);
+    Statement body1 = body.accept(shuttle);
+    List<CatchBlock> catchBlocks1 = new ArrayList<>();
+    for (CatchBlock cb: catchBlocks) {
+      Statement cbBody = cb.body.accept(shuttle);
+      catchBlocks1.add(
+          Expressions.catch_(cb.parameter, cbBody));
+    }
+    Statement fynally1 =
+        fynally == null ? null : fynally.accept(shuttle);
```
Comment (michaelmior) [inline linq4j/src/main/java/org/apache/calcite/linq4j/tree/TryStatement.java]: Looks like this should be `finally`?

File: linq4j/src/test/java/org/apache/calcite/linq4j/test/InlinerTest.java [hunk 2]
```diff
@@ -194,6 +196,40 @@ void checkAssignInConditionOptimizedOut(int modifiers, String s) {
...
+
+  @Test public void testInlineInTryCatchStatement() {
+    // final int t = 1;
+    // final int u;
+    // try {
+    //   u = t + 2;
+    // } catch (Exception e) {
+    //   throw e;
+    // }
+    // return u;
```
Comment (michaelmior) [inline linq4j/src/test/java/org/apache/calcite/linq4j/test/InlinerTest.java]: The commented out code above should be removed.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/ReflectiveCallNotNullImplementor.java [hunk 3]
```diff
@@ -47,17 +47,34 @@ public Expression implement(RexToLixTranslator translator,
...
+    }
+    return translator.handleMethodCheckedExceptions(callExpr);
   }
 
+  private boolean containsCheckedException(Method method) {
+    Class[] exceptions = method.getExceptionTypes();
+    if (exceptions == null || exceptions.length == 0) {
+      return false;
+    }
+    for (Class clazz: exceptions) {
```
Comment (michaelmior) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/ReflectiveCallNotNullImplementor.java]: Sorry this is getting really nitpicky, but there should be a space on either side of the `:`.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java [hunk 4]
```diff
@@ -622,6 +624,35 @@ public Expression handleNull(Expression input, RexImpTable.NullAs nullAs) {
...
+        callExpr.getType(), list.newName("method_call"));
+    list.add(Expressions.declare(Modifier.FINAL, methodCall, null));
+    Statement st = Expressions.statement(
+        Expressions.assign(methodCall, callExpr));
+    // Catch Block, wrap checked exception in unchecked exception
+    ParameterExpression e = Expressions.parameter(0, Exception.class, "e");
+    Expression uncheckedException =
+        Expressions.new_(RuntimeException.class, e);
+    CatchBlock cb = Expressions.catch_(e,
+        Expressions.throw_(uncheckedException));
```
Comment (michaelmior) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: The line breaks here seem unnecessary as these L648-651 could fit on two lines without being too long. (L644-645 could also be collapsed to a single line.)

File: core/src/test/java/org/apache/calcite/test/UdfTest.java [hunk 5]
```diff
@@ -219,6 +225,35 @@
     withUdf().query(sql).returns(expected);
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-3195">[CALCITE-3195]
+   * Handle UDF that throws checked exceptions in enumerable code generator</a>. */
```
Comment (michaelmior) [inline core/src/test/java/org/apache/calcite/test/UdfTest.java]: I suggest rewording to "Handle **a** UDF that throws in **the Enumerable** code generator"

File: linq4j/src/main/java/org/apache/calcite/linq4j/tree/TryStatement.java (hunk 1 above)
Comment (DonnyZone) [reply linq4j/src/main/java/org/apache/calcite/linq4j/tree/TryStatement.java]: @michaelmior Thanks for review! I addressed your comments. 
As "finally" is a keyword, I follow the variable name **fynally** in `TryStatement`.

Comment (michaelmior) [reply linq4j/src/main/java/org/apache/calcite/linq4j/tree/TryStatement.java]: Of course. This slipped my mind when reviewing :)

**Comment by michaelmior** [comment]
Thanks for the changes! Merging now.
